iap/iap.go

```diff
@@ -0,0 +1,515 @@
+	}
+}
+
+type InAppPurchaseDataHuawei struct {
+	ApplicationID string `json:"applicationId"`
+	AutoRenewing  bool   `json:"autoRenewing"`
+	OrderId       string `json:"orderId"`
+	Kind          int    `json:"kind"`
+	PackageName   string `json:"packageName"`
+	ProductId     string `json:"productId"`
+	PurchaseTime  int64  `json:"purchaseTime"`
+	PurchaseToken string `json:"purchaseToken"`
+	AccountFlag   int    `json:"accountFlag"`
+	PurchaseType  int    `json:"purchaseType"` // Omitted field in production, value set to 0 in sandbox env.
+}
+
+type ValidateReceiptHuaweiResponse struct {
+	ResponseCode      string                  `json:"responseCode"`
+	ResponseMessage   string                  `json:"responseMessage"`
+	PurchaseTokenData InAppPurchaseDataHuawei `json:"purchaseTokenData"`
+	DataSignature     string                  `json:"dataSignature"`
+}
+
+type AccessTokenHuawei struct {
+	// App-level access token.
+	AccessToken string `json:"access_token"`
+
+	// Remaining validity period of an access token, in seconds.
+	ExpiresIn int64 `json:"expires_in"`
+	// This value is always Bearer, indicating the type of the returned access token.
+	// TokenType    string	`json:"token_type"`
+
+	// Save the timestamp when AccessToken is obtained
+	ExpiredAt int64 `json:"-"`
+
+	// Request header string
+	HeaderString string `json:"-"`
+}
+
+func GetHuaweiAccessToken(ctx context.Context, httpc *http.Client, clientID, clientSecret string) (string, error) {
+	const authUrl = "https://oauth-login.cloud.huawei.com/oauth2/v3/token"
+
+	urlValue := url.Values{"grant_type": {"client_credentials"}, "client_id": {clientID}, "client_secret": {clientSecret}}
+	body := urlValue.Encode()
+	req, err := http.NewRequestWithContext(ctx, "POST", authUrl, strings.NewReader(body))
+	if err != nil {
+		return "", err
+	}
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+	req.Header.Set("Accept", "application/json")
+
+	resp, err := httpc.Do(req)
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	switch resp.StatusCode {
+	case 200:
+		buf, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return "", err
+		}
+
+		var out AccessTokenHuawei
+		if err := json.Unmarshal(buf, &out); err != nil {
+			return "", err
+		}
+		return out.AccessToken, nil
+	default:
+		return "", fmt.Errorf("non-200 response from Huawei auth: %+v", resp)
+	}
+}
+
+// Validate an IAP receipt with the Huawei API
+func ValidateReceiptHuawei(ctx context.Context, httpc *http.Client, pubKey, clientID, clientSecret, purchaseData, signature string) (*ValidateReceiptHuaweiResponse, *InAppPurchaseDataHuawei, []byte, error) {
+	if len(purchaseData) < 1 {
+		return nil, nil, []byte{}, errors.New("'purchaseData' must not be empty")
+	}
+
+	if len(signature) < 1 {
+		return nil, nil, []byte{}, errors.New("'signature' must not be empty")
+	}
+
+	data := &InAppPurchaseDataHuawei{PurchaseType: 2} // Set to placeholder value because field is omitted by prod purchase.
```

Comment:
As discussed use a sentinel/placeholder value outside any possible valid range. Perhaps `-1` would be best?